cmd/snap-bootstrap/hybrid_users_import.go

```diff
@@ -0,0 +1,582 @@
+			uid:    uid,
+			gid:    gid,
+			groups: userToGroups[name],
+			shell:  shell,
+
+			passwdEntry: entry,
+			shadowEntry: shadowEntries[name],
+		}
+
+		if !filter(u) {
+			continue
+		}
+
+		users[name] = u
+	}
+
+	return users, nil
+}
+
+type group struct {
+	name  string
+	gid   int
+	users []string
+
+	groupEntry   string
+	gshadowEntry string
+}
+
+// parseGroups parses groups from the given root directory, using the etc/group
+// and etc/gshadow files. The caller can provide a filter to omit some of the
+// returned groups by returning false from the given function.
+func parseGroups(root string, filter func(group) bool) (map[string]group, error) {
+	if filter == nil {
+		filter = func(group) bool { return true }
+	}
+
+	groupEntries, err := entriesByName(filepath.Join(root, "etc/group"))
+	if err != nil {
+		return nil, err
+	}
+
+	gshadowEntries, err := entriesByName(filepath.Join(root, "etc/gshadow"))
+	if err != nil {
+		return nil, err
+	}
+
+	groups := make(map[string]group)
+	for name, entry := range groupEntries {
+		parts := strings.Split(entry, ":")
+		if len(parts) != 4 {
+			logger.Noticef("skipping importing group with invalid entry: %s", entry)
+			continue
+		}
+
+		if name != parts[0] {
+			return nil, errors.New("internal error: group entry inconsistent with parsed data")
+		}
+
+		gid, err := strconv.Atoi(parts[2])
+		if err != nil {
+			logger.Noticef("skipping importing group %q with invalid gid: %v", name, err)
+			continue
+		}
+
+		var users []string
+		if len(parts[3]) > 0 {
+			users = strings.Split(parts[3], ",")
+		}
+
+		g := group{
+			name:         name,
+			gid:          gid,
+			users:        users,
+			groupEntry:   entry,
+			gshadowEntry: gshadowEntries[name],
+		}
+
+		if !filter(g) {
+			continue
+		}
+
+		groups[name] = g
+	}
+
+	return groups, nil
+}
+
+// parseGroupsForUser reads the contents of a file in the style of /etc/group
+// and returns a mapping of users in the file to the groups that each user is a
+// member of.
+func parseGroupsForUser(path string) (usersToGroups map[string][]string, err error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	usersToGroups = make(map[string][]string)
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "" {
+			continue
+		}
+
+		parts := strings.Split(line, ":")
+		if len(parts) != 4 {
+			continue
+		}
+
+		group := parts[0]
+		if parts[3] == "" {
+			continue
+		}
+
+		users := strings.Split(parts[3], ",")
+		for _, user := range users {
+			usersToGroups[user] = append(usersToGroups[user], group)
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return usersToGroups, nil
+}
+
+// parseShells parses a file in the format of /etc/shells and returns a list of
+// the valid login shells.
+func parseShells(root string) ([]string, error) {
+	f, err := os.Open(filepath.Join(root, "etc/shells"))
+	if err != nil {
+		return nil, err
```

Comment:
Should we return a fixed short allow list (/bin/bash and /bin/sh) here? Maybe we should not rely that much in having /etc/shells, as it is not such a fundamental file as passwd/groups (and the system might be somewhat broken so we should try to be resilient to that).

Reply:
Good idea, done!